uwblueprint/extend-a-family
# Facilitator's Learners GET + PUT Endpoints

## Notion ticket link

[Edit Facilitator’s Learners](https://www.notion.so/uwblueprintexecs/Edit-Facilitator-s-Learners-10a10f3fb1dc80c4bf43ec330c16345c?pvs=4)
[Get Facilitator’s Learners](https://www.notion.so/uwblueprintexecs/Get-Facilitator-s-Learners-10a10f3fb1dc80a9851ddf4f535925ea?pvs=4)



## Implementation description
- Added `GET: users/myLearners`
  - Added `isAuthorizedByRole` validator to verify user is faciliator
  - Use access token to determine facilitator userId

- Added `PUT: users/myLearner/:userId`
  - Added `isAuthorizedByRole` validator to verify user is faciliator
  - Added `updateUserDtoValidator` to verify the params
  - Use access token to determine facilitator userId
  - Verify selected userId corresponds to learner belonging to facilitator



## Steps to test
`GET: users/facilitator/myLearners`
1. Log into a facilitator account with at least one learner (see `Adding a Learner to a Facilitator` instructions below)
2. Use the EAF Dev tool chrome extension to get the auth token
   - OR go to Inspect Element > Application > Local storage > `http://localhost:3000` >`localhost:AUTHENTICATED_USER` > `authToken`
3. Go to Postman and send a request to `GET: users/facilitator/myLearners` with the Authorization token set to the auth token that was copied

Expected: 
- Receive an array of learners in the type of `UserDTO`s on Postman

`GET: users/facilitator/myLearners/:userId`
1. Log into a facilitator account with at least one learner (see `Adding a Learner to a Facilitator` instructions below)
2. Get the learner's `userID` from MongoDB
3. Use the EAF Dev tool chrome extension to get the auth token
   - OR go to Inspect Element > Application > Local storage > `http://localhost:3000` >`localhost:AUTHENTICATED_USER` > authToken
4. Go to Postman and send a request to `PUT: users/facilitator/:userID` with the Authorization token set to the auth token that was copied
   - Set the payload in the following format:
```
{
    "firstName": "Maggie2",
    "lastName": "Learner",
    "email": "maggiechen+learner@uwblueprint.org",
    "role": "Learner"
}
```
Expected: 
- Receive an learner with the updated information specified in the payload in the form of `UserDTO` on Postman
- Corresponding changes are made in MongoDB

**Adding a Learner to a Facilitator**
- Currently there is no way to register as a learner or an administrator, you can only register as a learner. So, we have to work around that to test this ticket.
1. Create another facilitator account
2. Go to MongoDB and find the facilitator account you just created
3. Change the facilitator account's role to Learner
4. Copy the `ObjectId` of the now learner account
5. In MongoDB, navigate to the facilitator account you wish to add this learner to
6. Add the copied learner account `ObjectId` to the learners array under this facilitator






## What should reviewers focus on?
* 


## Checklist
